beacon-chain/p2p/utils.go

```diff
@@ -123,43 +125,144 @@ func privKeyFromFile(path string) (*ecdsa.PrivateKey, error) {
 
 // Retrieves node p2p metadata from a set of configuration values
 // from the p2p service.
-// TODO: Figure out how to do a v1/v2 check.
+// When using static peer id, metaDataFromConfig returns default V0 metadata.
 func metaDataFromConfig(cfg *Config) (metadata.Metadata, error) {
-	defaultKeyPath := path.Join(cfg.DataDir, metaDataPath)
-	metaDataPath := cfg.MetaDataDir
+	// NOTE: Load V0 metadata by default because:
+	// - As the p2p service accesses metadata as an interface, and all versions implement the interface,
+	//   there is no error in calling the fields of higher versions. It just returns the default value.
+	// - This approach allows us to avoid unnecessary code changes when the metadata version bumps.
+	// - `RefreshPersistentSubnets` runs twice every slot and it manages updating and saving metadata.
+	defaultMd := &pb.MetaDataV0{
+		SeqNumber: 0,
+		Attnets:   bitfield.NewBitvector64(),
+	}
+	wrappedDefaultMd := wrapper.WrappedMetadataV0(defaultMd)
 
-	_, err := os.Stat(defaultKeyPath)
-	defaultMetadataExist := !os.IsNotExist(err)
-	if err != nil && defaultMetadataExist {
+	// Return default metadata for initialization if
+	// 1. Node is not using static peer ID, and
+	// 2. Fulu is not enabled.
+	if !cfg.StaticPeerID && !params.FuluEnabled() {
+		return wrappedDefaultMd, nil
+	}
+
+	mdPath, exist, err := resolveMetaDataPath(cfg)
+	if err != nil {
 		return nil, err
 	}
-	if metaDataPath == "" && !defaultMetadataExist {
-		metaData := &pb.MetaDataV0{
-			SeqNumber: 0,
-			Attnets:   bitfield.NewBitvector64(),
-		}
-		dst, err := proto.Marshal(metaData)
+
+	if exist {
+		md, err := metaDataFromFile(mdPath)
 		if err != nil {
+			if errors.Is(err, errUnexpectedMetadataSize) {
+				// In case previous metadata file is encoded by proto,
+				// we need to migrate it into ssz encoded version.
+				return migrateFromProtoToSsz(mdPath)
+			}
 			return nil, err
 		}
-		if err := file.WriteFile(defaultKeyPath, dst); err != nil {
-			return nil, err
+		return md, err
+	}
+	if err := saveMetaDataToFile(mdPath, wrappedDefaultMd); err != nil {
+		return nil, err
+	}
+
+	return wrappedDefaultMd, nil
+}
+
+// resolveMetaDataPath returns path and the existence of that path.
+func resolveMetaDataPath(cfg *Config) (string, bool, error) {
+	mdPath := cfg.MetaDataDir
+	if mdPath == "" {
+		mdPath = path.Join(cfg.DataDir, metaDataPath)
+	}
+
+	// Return path and existence of the file.
+	dirExists, err := file.Exists(mdPath, file.Regular)
+	if err != nil {
+		return mdPath, false, err
+	}
+	return mdPath, dirExists, nil
+}
+
+// metaDataFromFile retrieves unmarshalled p2p metadata from file.
+func metaDataFromFile(path string) (metadata.Metadata, error) {
```

Comment:
I think finding the version of the Metadata based on the size of the SSZ-encoding struct is quite weak.
We could have 2 different versions with the same size.